### PR TITLE
[WIP] New post-threshold Slime mutation: 'Mitosis'

### PR DESCRIPTION
**Pull request by harison86**
First time contributing, my follow-up to the suggestion I made here: http://smf.cataclysmdda.com/index.php?topic=15077.0

The primary purpose of Mitosis as a whole is to give Slime mutants a sidegrade to slimesprings that can ride in vehicles and be given orders. Secondary purposes include buffing the Slime mutation line and just being cool.
Mitosis has Omnicellular as a requirement, but Omnicellular does not turn into Mitosis.

In effect, Mitosis allows the player to spawn NPCs intended to be akin to replicas of a post-threshold Slime mutant. The class used by the Slimelings spawned has most of the Slime mutations, Illiterate, Fragile, no skills (except for some Bartering skill that seems to be inherent to all NPCs) and -8 strength bonus. They do not have Omnicellular in order to prevent them wasting food and water, nor Robust Genetics to prevent them from gaining strength over time from positive mutations.

Slimeling NPCs have 0 base strength on average, ranging from -4 to 4, rarely going to those extremes. At 0 strength, a Slimeling can carry 17.2 lbs and has 16 HP on each limb. 17.2 lbs is just enough capacity for a light survivor suit, light survivor hood, and a weapon. They have 6 health pools but can be one-shot by strong enemies such as brutes. This hopefully brings them closer to a balanced state as a technically infinitely spawnable NPC, and can be explained as being made primarily of slime with even less structural integrity than the player. The player is limited by the immediate cost of 100 nutrition and thirst, and the Slimelings requiring twice as much food and water as other NPCs if needs are enabled/when needs can be handled more smoothly.

I've tested these changes and they compile and work correctly as of build 6851. Minor issues include Slimelings receiving full epilogues when the player dies, base stat changes from mutations not applying to NPCs, and Slimelings cursing like a drunken sailor if the player is in their way. That's no way to talk to your parent.

### CODE REVIEW

**Comment by Coolthulhu** [comment]
That hp thing is pretty bad. This means they will need to be sealed in armor to be usable, which is the opposite of what should be happening. Armor upkeep is really tedious with NPCs, especially considering you can't sew NPC armor without removing it or order them to sew.
Instead, they should have more HP, but a mutation that prevents wearing armor.

They have bendy type mutations, but also flat -8 strength. Those stack. NPCs can't have negative strength, so they will probably always have 0. It would be better to just specify strength as -100 if this is desired.

Slimy clones should not be intelligent. Would be good if they always had 0 intelligence.

File: src/mutation.cpp [hunk 1]
```diff
@@ -481,6 +482,41 @@ void player::activate_mutation( const trait_id &mut )
         }
         tdata.powered = false;
         return;
+    } else if( mut == trait_SLIMESPAWNER2 ) {
```
Comment (Coolthulhu) [inline src/mutation.cpp]: This whole block is copypasted section above. This is totally wrong.
Two options:
* Extract a function and call it here and above
* `} else if( mut == trait_SLIMESPAWNER || mut == trait_SLIMESPAWNER2 ) {` and then later pick summon monster or summon npc based on `mut`

File: src/mutation.cpp [hunk 2]
```diff
@@ -458,61 +458,37 @@ void player::activate_mutation( const trait_id &mut )
         }
         // Oops, no room to divide!
         if (valid.size() == 0) {
-            add_msg_if_player(m_bad, _("You focus, but are too hemmed in to birth a new slimespring!"));
+            if ( mut == trait_SLIMESPAWNER ) {
+                add_msg_if_player(m_bad, _("You focus, but are too hemmed in to birth a new slimespring!"));
+            } else {
+                add_msg_if_player(m_bad, _("You focus, but are too hemmed in to birth a new slimeling!"));
+            }
```
Comment (Night-Pryanik) [inline src/mutation.cpp]: Why this duplicate messages? If I'm not mistaken, player can't have `SLIMESPAWNER` and `SLIMESPAWNER2` at the same time? If so, remove the if-else check and show the message for either of the mutations.

Comment (harison86) [reply src/mutation.cpp]: They can have both at the same time. `SLIMESPAWNER` is a requirement for `SLIMESPAWNER2`, but `SLIMESPAWNER` does not turn into `SLIMESPAWNER2`.

**Comment by kevingranade** [comment]
> Instead, they should have more HP, but a mutation that prevents wearing armor.

I agree with this, making them simple and less demanding for the player to interact with is a major plus.

File: data/json/mutations.json [hunk 3]
```diff
@@ -4167,6 +4168,23 @@
...
+        "visibility" : 0,
+        "ugliness" : 0,
+        "description" : "Your body has changed further to allow you to create imperfect replicas of yourself. They are obedient and more useful than previous iterations, but also more resource intensive.",
+        "valid" : false,
+        "purifiable" : false,
+        "prereqs" : ["SLIMESPAWNER"],
+        "threshreq" : ["THRESH_SLIME"],
+        "category" : ["MUTCAT_SLIME"],
+        "active"     :    true,
+        "cost"       :    100,
```
Comment (Coolthulhu) [inline data/json/mutations.json]: That cost is pretty low if we assume the created critter is roughly human-sized.
Thirst above 260 prevents using mutations and above 1200 causes death.

File: data/json/mutations.json [hunk 4]
```diff
@@ -4167,6 +4168,23 @@
         "thirst"     :    true
     },{
         "type" : "mutation",
+        "id" : "SLIMESPAWNER2",
+        "name" : "Mitosis",
+        "points" : 10,
+        "visibility" : 0,
+        "ugliness" : 0,
+        "description" : "Your body has changed further to allow you to create imperfect replicas of yourself. They are obedient and more useful than previous iterations, but also more resource intensive.",
```
Comment (Coolthulhu) [inline data/json/mutations.json]: Descriptions should have 2 spaces between sentences.

File: data/json/npcs/classes.json [hunk 5]
```diff
@@ -421,5 +421,50 @@
...
+            [ "DISIMMUNE", 100 ],
+            [ "PARAIMMUNE", 100 ],
+            [ "POISONOUS", 100 ],
+            [ "SLIMY", 100 ],
+            [ "SLIME_HANDS", 100 ],
+            [ "DEX_UP", 100 ],
+            [ "INT_UP", 100 ],
+            [ "SMELLY2", 100 ],
+            [ "DEFORMED3", 100 ],
+            [ "HOLLOW_BONES", 100 ],
```
Comment (Coolthulhu) [inline data/json/npcs/classes.json]: Hollow bones will make them go fast.

Comment (harison86) [reply data/json/npcs/classes.json]: Is there a particular reason they shouldn't go fast? They will be able to attack faster, however this is balanced out by their frailty. If the concern is them running ahead of or away from the player into danger, the Slimeling's increased movement speed does not appear to be detrimental in that regard; they can be ordered not to fight and the player can still reach enemies first with the Slimeling's following a few spaces behind.

**Comment by Coolthulhu** [comment]
They can still wear XL armor just fine, which defeats the point of them not having an armor. This will probably require some hard code to make them totally unable to wear any armor. Probably just one check in one of the armor wearing functions.

I'd also drop that unstable genetics mutation - it makes them more attention-heavy, since you need to know what mutations they have if you want to optimize them. Also the food mutations - if anything, they should be easy to keep fed since that's tedious.

There are some problems not added here, but that impair the current PR rather heavily:
* All slimelings are named "Slimeling". There is no way to tell them apart. Even just adding a number after their name would be good (though some proper naming scheme with thematic names would be cooler)
* The code for adding NPC traits does not set them as base traits. `set_mutation` sets them as acquired mutations, but we want traits. Currently the only way to set a trait is `toggle_trait` (when the trait is unset). The differences are small until you start using mutagen, but with mutagen they start showing.
* Attacking your own slimelings can make your faction permanently hostile to you. This will make all future slimelings spawn hostile.

Also some tiny details, fixing those is unnecessary, but they're still worth noting:
* Slimelings are always male (should copy player's sex)
* Slimelings leave normal human corpses which can be butchered for 5 human meats. Butcher two, add a pinch of salt and a charcoal smoker and you get a brat bologna (320 nutrition). If you have the digestion bionic, you can gain up to 80 nutrition (not counting other edible human parts) per "suicide" that way.
* Slimelings don't really like you that much - they're just friends, nothing more. This is visible in "teach me" chances.

File: src/game.h [hunk 6]
```diff
@@ -312,6 +314,10 @@ class game
          * Returns a `nullptr` if the monster could not be created.
          */
         monster *summon_mon( const mtype_id& id, const tripoint &p );
+        /**
+         * Summons an NPC.
```
Comment (BevapDin) [inline src/game.h]: A function named "summon_npc" summons an NPC? That documentation is a pointless wast of time to read. The function name tells exactly the same. Either remove it completely or make it into something that *adds* information.

File: src/game.h [hunk 7]
```diff
@@ -10,6 +10,7 @@
 #include "item_location.h"
 #include "cursesdef.h"
 #include "ranged.h"
+#include "npc.h"
```
Comment (BevapDin) [inline src/game.h]: You don't need that included. A forward declaration of `npc_template` is enough.

File: src/game.cpp [hunk 8]
```diff
@@ -6626,6 +6626,18 @@ monster *game::summon_mon( const mtype_id& id, const tripoint &p )
...
+void *game::summon_npc( const string_id<npc_template> &id, const tripoint &p )
+{
+    std::shared_ptr<npc> temp = std::make_shared<npc>();
+    temp->load_npc_template( id );
+    temp->normalize();
+    temp->spawn_at_precise( { get_levx(), get_levy() }, p );
+    overmap_buffer.insert_npc( temp );
+    temp->mission = NPC_MISSION_NULL;
+    load_npcs();
+    return nullptr;
```
Comment (BevapDin) [inline src/game.cpp]: ??? What? Why? And why does the function return `void*` anyway?

**Pull request closed without merging**